src/support/hazard_weak.c

```diff
@@ -0,0 +1,277 @@
+/*-
+ * Copyright (c) 2014-present MongoDB, Inc.
+ * Copyright (c) 2008-2014 WiredTiger, Inc.
+ *	All rights reserved.
+ *
+ * See the file LICENSE for redistribution information.
+ */
+
+#include "wt_internal.h"
+
+#define WT_HAZARD_WEAK_FORALL(s, wha, whp)                     \
+    for (wha = (s)->hazard_weak; wha != NULL; wha = wha->next) \
+        for (whp = wha->hazard; whp < wha->hazard + wha->hazard_inuse; whp++)
+
+#define WT_HAZARD_WEAK_FORALL_BARRIER(s, wha, whp)               \
+    for (wha = (s)->hazard_weak; wha != NULL; wha = wha->next) { \
+        uint32_t __hazard_inuse;                                 \
+        WT_ORDERED_READ(__hazard_inuse, wha->hazard_inuse);      \
+        for (whp = wha->hazard; whp < wha->hazard + __hazard_inuse; whp++)
+
+#define WT_HAZARD_WEAK_FORALL_BARRIER_END }
+
+/*
+ * __wt_hazard_weak_close --
+ *     Verify that no weak hazard pointers are set.
+ */
+void
+__wt_hazard_weak_close(WT_SESSION_IMPL *session)
+{
+    WT_HAZARD_WEAK *whp;
+    WT_HAZARD_WEAK_ARRAY *wha;
+    uint32_t nhazard_weak;
+    bool found;
+
+    /*
+     * Check for a set weak hazard pointer and complain if we find one. We could just check the
+     * session's weak hazard pointer count, but this is a useful diagnostic.
+     */
+    for (found = false, nhazard_weak = 0, wha = session->hazard_weak; wha != NULL;
+         nhazard_weak += wha->nhazard, wha = wha->next)
+        for (whp = wha->hazard; whp < wha->hazard + wha->hazard_inuse; whp++)
+            if (whp->ref != NULL) {
+                found = true;
+                break;
+            }
+
+    if (nhazard_weak == 0 && !found)
+        return;
+
+    __wt_errx(
+      session, "session %p: close weak hazard pointer table: table not empty", (void *)session);
+
+    WT_HAZARD_WEAK_FORALL (session, wha, whp)
+        if (whp->ref != NULL) {
+            whp->ref = NULL;
+            --wha->nhazard;
+            --nhazard_weak;
+        }
+
+    if (nhazard_weak != 0)
+        __wt_errx(session,
+          "session %p: close weak hazard pointer table: count didn't match entries",
+          (void *)session);
+}
+
+/*
+ * hazard_weak_grow --
+ *     Grow a weak hazard pointer array.
```

Comment:
Could we add a clearer/more detailed description of what happens in this function? It's easy to miss what is going on here.

Reply:
Added.